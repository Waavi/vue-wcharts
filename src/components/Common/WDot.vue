<template>
    <WTrans
        :initialProps="{
            r: 0
        }"
        :transition="transition"
    >
        <circle
            :cx="x"
            :cy="y"
            :stroke="styles.stroke"
            :fill="styles.fill"
            :r="rStyle"
            :stroke-width="styles.strokeWidth"
            :style="{ opacity: styles.opacity, cursor: styles.cursor }"
            @click="$emit('onClick', info)"
            @mouseenter="handleMouseEnter"
            @mouseleave="Chart.cleanActive"
        />
    </WTrans>
</template>

<script>
import VueTypes from 'vue-types'
import { WTrans } from '../../transitions'

export default {
    name: 'WDot',
    type: 'cartesian',
    inject: ['Chart'],
    components: {
        WTrans,
    },
    props: {
        index: VueTypes.number.isRequired,
        cartesianIndex: VueTypes.number.isRequired,
        x: VueTypes.number.isRequired,
        y: VueTypes.number.isRequired,
        styles: VueTypes.shape({
            fill: VueTypes.string,
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            radius: VueTypes.number,
            hoverRadius: VueTypes.number,
            opacity: VueTypes.number,
            cursor: VueTypes.string.def('default'),
        }).isRequired,
        transition: VueTypes.string.isRequired,
        info: VueTypes.shape({
            id: VueTypes.any,
            label: VueTypes.any,
            value: VueTypes.array,
        }).loose,
    },
    computed: {
        // Return styles active or default point,
        rStyle () {
            if (!this.Chart.active.el) return this.styles.radius
            const { el } = this.Chart.active
            return el.id === this.cartesianIndex && el.point === this.index ? this.styles.hoverRadius : this.styles.radius
        },
    },
    methods: {
        // Set active element
        handleMouseEnter (event) {
            const { setActive } = this.Chart
            setActive(
                this.info,
                event
            )
        },
    },
}
</script>
