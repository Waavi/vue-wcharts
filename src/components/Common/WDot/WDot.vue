<template>
    <WTrans
        :initialProps="{
            r: 0
        }"
        :transition="transition"
    >
        <circle
            :cx="x"
            :cy="y"
            :stroke="stroke"
            :stroke-width="strokeWidth || themeStyles.strokeWidth"
            :fill="fill"
            :r="rStyle"
            :style="dotStyles"
            v-on="dotListeners"
        />
    </WTrans>
</template>

<script>
import VueTypes from 'vue-types'
import merge from '../../../utils/merge'
import { WTrans } from '../../../transitions'
import themeMixin from '../../../mixins/theme'

export default {
    name: 'WDot',
    type: 'cartesian',
    inject: ['Chart'],
    components: {
        WTrans,
    },
    mixins: [themeMixin],
    props: {
        index: VueTypes.number.isRequired,
        cartesianIndex: VueTypes.number.isRequired,
        trigger: VueTypes.oneOf(['hover', 'click', 'manual']).def('hover'),
        x: VueTypes.number.isRequired,
        y: VueTypes.number.isRequired,
        fill: VueTypes.string,
        stroke: VueTypes.string,
        strokeWidth: VueTypes.number,
        radius: VueTypes.number,
        hoverRadius: VueTypes.number,
        info: VueTypes.shape({
            id: VueTypes.any,
            label: VueTypes.any,
            value: VueTypes.array,
        }).loose,
        styles: VueTypes.object.def({}),
        transition: VueTypes.string.isRequired,
    },
    computed: {
        dotStyles () {
            return {
                ...this.themeStyles.styles,
                ...this.styles,
                ...(this.trigger === 'click' ? { cursor: 'pointer' } : {}),
            }
        },
        // Event Listeners
        dotListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    if (this.trigger === 'click') this.handleActive(event)
                    this.$emit('onClick', this.info)
                },
                mouseenter: (event) => {
                    if (this.trigger === 'hover') this.handleActive(event)
                    this.$emit('onMouseenter', this.info)
                },
                mouseleave: () => {
                    if (['hover', 'click'].includes(this.trigger)) this.Chart.cleanActive()
                    this.$emit('onMouseleave')
                },
            })
        },
        // Return styles active or default point,
        rStyle () {
            if (!this.Chart.active.el) return this.radius
            const { el } = this.Chart.active
            return el.id === this.cartesianIndex && el.point === this.index ? (this.hoverRadius || this.themeStyles.hoverRadius) : (this.radius || this.themeStyles.radius)
        },
    },
    methods: {
        // Set active element
        handleActive (event) {
            const { setActive } = this.Chart
            setActive(
                this.info,
                event
            )
        },
    },
}
</script>
