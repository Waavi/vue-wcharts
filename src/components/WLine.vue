<template>
    <path
        ref="line"
        :d="linePath"
        stroke="black"
        fill="none"
        :strokeWidth="3"
    />
</template>

<script>
import VueTypes from 'vue-types'
import { line as d3Line } from 'd3'

export default {
    name: 'WLine',
    type: 'cartesian',
    inject: ['Cartesian'],
    props: {
        datakey: VueTypes.string.isRequired,
    },
    computed: {
        lineData () {
            return this.Cartesian.dataset.map((item, index) => ({
                x: index + 1,
                y: item[this.datakey],
            }))
        },
        defaultLine () {
            return d3Line()
                .x(d => this.Cartesian.xScale(d.x))
                .y(d => this.Cartesian.yScale(d.y))
        },
        linePath () {
            return this.defaultLine(this.lineData)
        },
    },
}
</script>
