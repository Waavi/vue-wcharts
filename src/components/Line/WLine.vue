<template>
    <g v-if="active">
        <WSpread
            axis="x"
            :transition="transition"
        >
            <path
                :d="linePath"
                :style="{ transition }"
                v-bind="stylesCmp"
                fill="none"
                v-on="lineListeners"
            />
            <defs
                v-if="area && stylesCmp.fill === ''"
            >
                <linearGradient :id="`areaGradient${_uid}`">
                    <stop
                        :stop-color="stylesCmp.stroke"
                        stop-opacity="0.5"
                    />
                </linearGradient>
            </defs>
            <path
                v-if="area"
                :d="areaPath"
                :fill="stylesCmp.fill || `url(#areaGradient${_uid})`"
                :style="{ transition }"
                v-on="lineListeners"
            />
        </WSpread>
        <g
            v-if="dot"
        >
            <template
                v-for="(dotItem, id) in dotsData"
            >
                <slot
                    name="dot"
                    :dot="dotItem"
                    :first="id === 0"
                    :last="id === dotsData.length - 1"
                    :Chart="Chart"
                    :transition="transition"
                >
                    <WDot
                        :key="`dot${dotItem.cartesianIndex}${dotItem.index}`"
                        v-bind="dotItem"
                        :transition="`all 250ms ${transEffect}`"
                        :trigger="trigger"
                        @onClick="$emit('onClickDot', $event)"
                    />
                </slot>
            </template>
        </g>
    </g>
</template>

<script>
import VueTypes from 'vue-types'
import d3Line from 'd3-shape/src/line'
import d3Area from 'd3-shape/src/area'
import cardinal from 'd3-shape/src/curve/cardinal'
import merge from '../../utils/merge'
import { WDot } from '../Common'
import animationMixin from '../../mixins/animation'
import themeMixin from '../../mixins/theme'
import visibleMixin from '../../mixins/visible'
import { WSpread } from '../../transitions'
import { isFunc, noNilInArray } from '../../utils/checks'

export default {
    name: 'WLine',
    type: 'cartesian',
    components: {
        WDot,
        WSpread,
    },
    mixins: [
        themeMixin,
        animationMixin,
        visibleMixin,
    ],
    inject: ['Chart'],
    props: {
        index: VueTypes.number.isRequired, // internal props set by the parent (WCartesian)
        datakey: VueTypes.string.isRequired,
        trigger: VueTypes.oneOf(['hover', 'click', 'manual']).def('click'),
        legend: VueTypes.string,
        curve: VueTypes.oneOfType([VueTypes.bool, VueTypes.func]).def(false),
        area: VueTypes.oneOfType([VueTypes.bool, VueTypes.string]).def(false),
        styles: VueTypes.shape({
            fill: VueTypes.string,
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            strokeDasharray: VueTypes.string,
        }).loose.def({}),
        dot: VueTypes.bool.def(false),
        dotStyles: VueTypes.shape({
            fill: VueTypes.string,
            stroke: VueTypes.string,
            strokeWidth: VueTypes.number,
            radius: VueTypes.number,
            hoverRadius: VueTypes.number,
        }).def({}),
        continued: VueTypes.bool.def(false),
    },
    // It's called by parent components to necessary calcs before be rendering
    // Componen is not mounted and cannot access to default props
    preload ({ parent, props, index }) {
        const { snap } = parent
        const { datakey } = props

        // Set datakeys by index
        snap.linesByIndex = { ...snap.linesByIndex, [index]: { datakey } }
    },
    computed: {
        active () {
            return this.Chart.activeElements.includes(this.index)
        },
        lineData () {
            return this.Chart.data
                .map((item, index) => item[this.datakey])
                .map((y, x) => (y ? [this.Chart.xScale(x), this.Chart.yScale(y)] : [null]))
        },
        dotsData () {
            if (!this.dot) return []
            const { data, axis } = this.Chart

            return data.reduce((acc, item, index) => {
                const [x, y] = this.lineData[index] || []
                if (y === undefined) return acc

                const info = {
                    id: this.index,
                    data: item,
                    point: index,
                    label: item[axis.x.datakey],
                    value: [{
                        value: item[this.datakey],
                        color: this.fillColor,
                    }],
                }

                return [
                    ...acc,
                    {
                        index,
                        cartesianIndex: this.index,
                        value: item[this.datakey],
                        x,
                        y,
                        info,
                        ...this.dotStylesCmp,
                    },
                ]
            }, [])
        },
        linePath () {
            const draw = d3Line().defined(noNilInArray)
            const data = this.continued ? this.lineData.filter(noNilInArray) : this.lineData

            if (this.curve) {
                draw.curve(isFunc(this.curve) ? this.curve : cardinal)
            }

            return draw(data)
        },
        areaPath () {
            if (!this.area) return null
            const draw = d3Area().defined(noNilInArray).y0(this.y0Area)
            const data = this.continued ? this.lineData.filter(noNilInArray) : this.lineData

            if (this.curve) {
                draw.curve(isFunc(this.curve) ? this.curve : cardinal)
            }

            return draw(data)
        },
        // Event Listeners
        lineListeners () {
            return merge({}, this.$listeners, {
                click: (event) => {
                    this.$emit('onClick', event)
                },
                mouseenter: (event) => {
                    this.$emit('onMouseenter', event)
                },
                mouseleave: () => {
                    this.$emit('onMouseleave')
                },
            })
        },
        // Styles
        stylesCmp () {
            const stroke = ((this.themeStyles || {}).styles || {}).stroke || this.styles.stroke || this.fillColor
            return {
                ...this.themeStyles.styles,
                ...this.styles,
                stroke,
            }
        },
        dotStylesCmp () {
            const styles = { ...this.themeStyles.dot, ...this.dotStyles }
            return {
                ...styles,
                stroke: styles.stroke || this.fillColor,
                fill: styles.fill || this.fillColor,
            }
        },
        fillColor () {
            return (this.colors || [])[this.index] || this.Chart.colors[this.index]
        },
    },
    methods: {
        y0Area (d, index) {
            if (typeof this.area !== 'string') return this.Chart.canvas.y1
            return this.Chart.data.map(item => this.Chart.yScale(item[this.area] || 0))[index]
        },
    },
}
</script>
