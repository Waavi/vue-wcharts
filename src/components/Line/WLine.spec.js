import { shallowMount, mount } from '@vue/test-utils'
import { curveStep } from 'd3-shape'
import WLine from './WLine.vue'
import WDot from '../Common/WDot/WDot.vue'

describe('Components/WLine', () => {
    const dataset = [
        {
            name: 'Page A', one: 4000, two: -2400, three: 2300, four: 1200, five: 2300,
        },
        {
            name: 'Page B', one: 3000, two: 2210, three: 1398, four: 1200, five: 1398,
        },
        {
            name: 'Page C', one: 9800, two: 2200, three: 0, four: 1200, five: 0,
        },
    ]

    const propsData = {
        index: 0,
        datakey: 'one',
        styles: {
            stroke: 'FFF',
        },
    }

    const provide = {
        Chart: {
            colors: ['red', 'red', 'red', 'red', 'red', 'red', 'red'],
            legends: ['One Bar', 'Two Bar'],
            space: [20, 20, 20, 20],
            activeElements: [0],
            canvas: {
                x0: 50, y0: 25, x1: 500, y1: 250,
            },
            dataset,
            data: dataset,
            xScale: a => a,
            yScale: a => a,
            setActive: () => undefined,
        },
    }

    const dotConfig = {
        propsData: { ...propsData, dot: true, dotStyles: { stroke: undefined, fill: undefined } },
        provide: {
            ...provide,
            Chart: {
                ...provide.Chart,
                datakeys: ['one', 'two', 'three'],
                colors: ['red', 'red', 'red'],
                axis: {
                    x: {
                        datakey: 'name',
                        name: '',
                    },
                },
                active: {
                    el: null,
                },
            },
        },
    }

    const defaultConfig = {
        propsData,
        provide,
    }

    it(`Should be render correctly`, () => {
        const wrapper = shallowMount(WLine, defaultConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with dots`, () => {
        const wrapper = shallowMount(WLine, dotConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Shouldn't be render`, () => {
        const wrapper = shallowMount(WLine, { ...defaultConfig, propsData: { ...propsData, visible: false } })
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with 3 dots`, () => {
        const wrapper = shallowMount(WLine, dotConfig)
        expect(wrapper.findAll(WDot)).toHaveLength(3)
    })

    it(`Should be render correctly with area`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, area: true } }
        const wrapper = shallowMount(WLine, customConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly with custom color area `, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, area: true, styles: { fill: 'red' } } }
        const wrapper = shallowMount(WLine, customConfig)
        expect(wrapper.html()).toMatchSnapshot()
    })

    it(`Should be render correctly without dot`, () => {
        const wrapper = shallowMount(WLine, defaultConfig)
        expect(wrapper.vm.dotsData).toEqual([])
    })

    it(`Should be render correctly with custom fillColor`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, index: 1, colors: ['#000', '#FFF'] } }
        const wrapper = shallowMount(WLine, customConfig)
        expect(wrapper.vm.fillColor).toEqual('#FFF')

        const customConfigTwo = { ...defaultConfig, propsData: { ...defaultConfig.propsData, index: 3, colors: undefined } }
        const wrapperTwo = shallowMount(WLine, customConfigTwo)
        expect(wrapperTwo.vm.fillColor).toEqual('#e8e8e8')
    })

    it(`Should be render correctly with curve and bool`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, area: true, curve: true } }
        const wrapper = shallowMount(WLine, customConfig)
        expect(wrapper.vm.linePath).toMatchSnapshot()
        expect(wrapper.vm.areaPath).toMatchSnapshot()
    })

    it(`Should be render correctly with curve and func`, () => {
        const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, area: true, curve: curveStep } }
        const wrapper = shallowMount(WLine, customConfig)
        expect(wrapper.vm.linePath).toMatchSnapshot()
        expect(wrapper.vm.areaPath).toMatchSnapshot()
    })

    describe('Events', () => {
        it('It emits the handleMouseEnter event', () => {
            const wrapper = mount(WLine, defaultConfig)
            wrapper.find('path').trigger('click')
            expect(wrapper.emitted('onClick')).toHaveLength(1)
        })
        it('It emits the handleMouseEnter event', () => {
            const wrapper = mount(WLine, defaultConfig)
            wrapper.find('path').trigger('mouseenter')
            expect(wrapper.emitted('onMouseenter')).toHaveLength(1)
        })
        it('It emits the mouseleave event', () => {
            const wrapper = mount(WLine, defaultConfig)
            wrapper.find('path').trigger('mouseleave')
            expect(wrapper.emitted('onMouseleave')).toHaveLength(1)
        })
        it('It emits the onClick event', () => {
            const wrapper = mount(WLine, dotConfig)
            wrapper.find(WDot).trigger('click')
            expect(wrapper.emitted('onClickDot')).toHaveLength(1)
        })
    })

    describe('Methods', () => {
        it('Should be return y0 coord on y0Area executed', () => {
            const customConfig = { ...defaultConfig, propsData: { ...defaultConfig.propsData, area: 'two', curve: true } }
            const wrapper = shallowMount(WLine, customConfig)
            expect(wrapper.vm.y0Area(null, 1)).toEqual(2210)
        })
    })
})
