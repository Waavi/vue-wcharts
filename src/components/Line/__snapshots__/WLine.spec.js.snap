// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/WLine Should be render correctly 1`] = `
<g>
  <WSpread-stub axis="x" transition="all 1s ease">
    <path d="M0,4000L1,3000L2,9800" fill="none" stroke="FFF" strokeWidth="2" strokeDasharray="0" style="transition: all 1s ease;"></path>
    <!---->
    <!---->
  </WSpread-stub>
  <!---->
</g>
`;

exports[`Components/WLine Should be render correctly with area 1`] = `
<g>
  <WSpread-stub axis="x" transition="all 1s ease">
    <path d="M0,4000L1,3000L2,9800" fill="none" stroke="FFF" strokeWidth="2" strokeDasharray="0" style="transition: all 1s ease;"></path>
    <defs>
      <linearGradient id="areaGradient15">
        <stop stop-color="FFF" stop-opacity="0.5"></stop>
      </linearGradient>
    </defs>
    <path d="M0,4000L1,3000L2,9800L2,250L1,250L0,250Z" fill="url(#areaGradient15)" style="transition: all 1s ease;"></path>
  </WSpread-stub>
  <!---->
</g>
`;

exports[`Components/WLine Should be render correctly with custom color area  1`] = `
<g>
  <WSpread-stub axis="x" transition="all 1s ease">
    <path d="M0,4000L1,3000L2,9800" fill="none" stroke="red" strokeWidth="2" strokeDasharray="0" style="transition: all 1s ease;"></path>
    <!---->
    <path d="M0,4000L1,3000L2,9800L2,250L1,250L0,250Z" fill="red" style="transition: all 1s ease;"></path>
  </WSpread-stub>
  <!---->
</g>
`;

exports[`Components/WLine Should be render correctly with dots 1`] = `
<g>
  <WSpread-stub axis="x" transition="all 1s ease">
    <path d="M0,4000L1,3000L2,9800" fill="none" stroke="FFF" strokeWidth="2" strokeDasharray="0" style="transition: all 1s ease;"></path>
    <!---->
    <!---->
  </WSpread-stub>
  <g>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="0" cartesianIndex="0" trigger="hover" x="0" y="4000" fill="white" stroke="red" strokeWidth="2" radius="4" hoverRadius="8" info="[object Object]" styles="[object Object]" transition="all 250ms ease" value="4000"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="1" cartesianIndex="0" trigger="hover" x="1" y="3000" fill="white" stroke="red" strokeWidth="2" radius="4" hoverRadius="8" info="[object Object]" styles="[object Object]" transition="all 250ms ease" value="3000"></WDot-stub>
    <WDot-stub colors="#48c0b6,#9d6af5,#ffde69,#e8e8e8,#54ddd9,#6e19ff,#dfbd46" index="2" cartesianIndex="0" trigger="hover" x="2" y="9800" fill="white" stroke="red" strokeWidth="2" radius="4" hoverRadius="8" info="[object Object]" styles="[object Object]" transition="all 250ms ease" value="9800"></WDot-stub>
  </g>
</g>
`;

exports[`Components/WLine Shouldn't be render 1`] = `undefined`;
